content/posts/2023-12-optimising-spark-memory-usage.md

```diff
@@ -0,0 +1,278 @@
+without any major changes in the implementation (like offloading the data from memory to the
+filesystem).
+
+## Plain data objects
+
+My colleague [Julian Gruber](https://github.com/juliangruber) discovered the first low-hanging
+fruit: apparently, plain data objects take more memory than class instances.
+
+Adding a step to map objects parsed from JSON to instances of a newly introduced `Measurement` class
+reduced our memory usage by 50%.
+
+The original version parsed measurements from JSON and then stored these objects for future
+evaluations.
+
+```js
+const measurements = /* fetch and parse JSON */
+const validMeasurements = measurements.map(measurement => {
+  try {
+    return {
+      ...measurement,
+      // additional fields parsed from data
+      }
+  } catch (err) {
+    logger.error('Invalid measurement:', err.message, measurement)
+    return null
+  }
+})
+```
+
+The new version converts each plain data object into a class instance. The class constructor
+performs preprocessing, e.g., parsing additional fields.
+
+```js
+const measurements = /* fetch and parse JSON */
+const validMeasurements = measurements.map(measurement => {
+  try {
+    return new Measurement(measurement)
+  } catch (err) {
+    logger.error('Invalid measurement:', err.message, measurement)
+    return null
+  }
+})
+```
+
+Take a look at the pull request
+[filecoin-station/spark-evaluate#84](https://github.com/filecoin-station/spark-evaluate/pull/84) if
+you are interested in more details.
+
+## Duplicate string values
+
+This first improvement sparked my interest. What other easy optimizations we can make?
+
+Let's take a look at an example measurement:
+
+```json
+{
+  "id": 280847321,
+  "spark_version": "1.6.0",
+  "zinnia_version": "0.15.0",
+  "participant_address": "0x000000000000000000000000000000000000dEaD",
+  "finished_at": "2023-12-11T08:31:00.157Z",
+  "timeout": false,
+  "start_at": "2023-12-11T08:30:29.947Z",
+  "status_code": 502,
+  "first_byte_at": null,
+  "end_at": "2023-12-11T08:30:59.954Z",
+  "byte_length": 0,
+  "attestation": null,
+  "inet_group": "uWJAEcUyzouo",
+  "car_too_large": false,
+  "cid": "bafybeice7venemgas2vijqc6xfdf6zequrxziajapm2j6hxckeu6fimjwa",
+  "provider_address": "/ip4/207.189.112.60/tcp/26197/http",
+  "protocol": "http"
+}
+```
+
+Setting aside timestamps, the following properties store a string value:
+
+- `spark_version` contains the version of the SPARK node that performed the retrieval.
+- `zinnia_version` is a field we ignore now; we don't keep it in `Measurement` instances.
+- `participant_address` contains the Station's wallet address in ETH format.
+- `inet_group` is an anonymous id of the IPv4 /24 subnet from which the measurement was submitted.
+- `cid` is the identifier of the content that the node attempted to retrieve.
+- `provider_address` is the address of the Storage Provider.
+- `protocol` is either `http`, `bitswap`, or `graphsync`.
```

Comment:
I don't think the value description matters here. What about just saying that we have strings values where know some repeat a lot - and then give an example?

I'm suggesting this in order to improve signal vs noise - as a reader I want to learn primarily about the optimizations and not about SPARK/Meridian